### PR TITLE
Email 2.2

### PR DESCRIPTION
**Pull request by StorytellerCZ**
Implements new features as per: https://github.com/meteor/meteor/discussions/11547

 - [x] modernization
 - [x] Send out email with custom API
 - [x] Implement nodemailer's known hosts
 - [x] MAIL_URL can be used for known hosts

### CODE REVIEW

**Comment by StorytellerCZ** [comment]
@xet7 what do you think?

**Comment by harryadel** [comment]
Isn't it a good a time to remove the fibers stuff, convert to async and do a major bump?

**Comment by StorytellerCZ** [comment]
@harryadel Let's leave that for the next version. I think these changes are enough for now.

**Comment by diavrank** [comment]
This feature is awesome. I was wondering if it also consider to add more than one transport in order to contemplate this [feature request](https://github.com/meteor/meteor-feature-requests/issues/409).

**Comment by xet7** [comment]
> @xet7 what do you think?

I'll try some way way to test this. I tried:
- Using meteor fork of this PR, but Wekan is not yet upgraded to newest Meteor
- Making fork of only email package, but it seems other packages depend on `email@2.0.0` package

Maybe I could first build current Wekan bundle, and then replace email.js etc in that bundle, and run that bundle. I'll try that next.

**Comment by StorytellerCZ** [comment]
@xet7 I would recommend just copying it into the local packages directory and then you can adjust the version as needed to make it work.

**Comment by xet7** [comment]
@StorytellerCZ 

> @xet7 I would recommend just copying it into the local packages directory and then you can adjust the version as needed to make it work.

I could not get it working in `packages` directory, so I copied code from `email.js` to `wekan/.build/bundle/programs/server/packages/email.js`.

Then I did some changes, to port it Node 12.x and Meteor 2.2 that Wekan currently used.

Here is changed email.js in .zip file: [email.zip](https://github.com/meteor/meteor/files/6872361/email.zip)

Here are changes I made:
```
export const
=>
const

const host = url?.split(':')[0];
=>
const host = url.split(':')[0];

  const transport = nodemailer.createTransport({
    service: settings?.service || service,
    auth: {
      user: settings?.user || user,
      pass: settings?.password || password
    }
  });
=>
  const transport = nodemailer.createTransport({
    service: settings.service || service,
    auth: {
      user: settings.user || user,
      pass: settings.password || password
    }
  });

const packageSettings = Meteor.settings.packages?.email || {};
=>
const packageSettings = Meteor.settings.packages.email || {};

if ((packageSettings?.service && knownHosts.includes(packageSettings)) || knownHosts.includes(url?.split(':')[0])) {
=>
if ((packageSettings.service && knownHosts.includes(packageSettings)) || knownHosts.includes(url.split(':')[0])) {

// New Meteor code
//const sendHooks = new Hook();
// Old Meteor code backport:
const sendHooks = [];

Email.hookSend = function (f) {
  // New Meteor code:
  //sendHooks.register(f);
  // Old Meteor code backport:
  sendHooks.push(f);
};
```

But now that I try to invite new user to Wekan board with user's email address, that would send email for invite, I get this error:
```
Exception while invoking method 'inviteUserToBoard' TypeError: Cannot read property 'send' of undefined
    at AccountsServer.Accounts.sendEnrollmentEmail (packages/accounts-password/password_server.js:804:9)
    at MethodInvocation.inviteUserToBoard (models/users.js:1196:18)
    at packages/check/match.js:118:15
    at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1234:12)
    at Object._failIfArgumentsAreNotAllChecked (packages/check/match.js:116:43)
    at maybeAuditArgumentChecks (packages/ddp-server/livedata_server.js:1768:18)
    at packages/ddp-server/livedata_server.js:719:19
    at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1234:12)
    at packages/ddp-server/livedata_server.js:717:46
    at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1234:12)
    at packages/ddp-server/livedata_server.js:715:46
    at new Promise (<anonymous>)
    at Session.method (packages/ddp-server/livedata_server.js:689:23)
    at packages/ddp-server/livedata_server.js:559:43
```

So I think some of the changes have not been ported correctly think.

Because I could not get this email package working yet with , I think next I'll try:

a) Figure out minimal changes to current Meteor 2.2 email package that would still work with Node 12.x, Meteor 2.2 and current Wekan.

b) Creating new example project with newest Meteor and this email package, to test does sending email work with newest Meteor.

**Comment by StorytellerCZ** [comment]
Or this might be related to the test failures that we see in CI.

**Comment by StorytellerCZ** [comment]
Alright! I have fixed the tests. The issue there was that they weren't properly cleaning after themselves.

@xet7 I think you will need to make changes in `package.js` as well to get it working. The package is now using the new syntax there which is why you are getting the undefined error.

**Comment by xet7** [comment]
@StorytellerCZ 

Thanks! I'll try newest changes.

**Comment by xet7** [comment]
@StorytellerCZ 

I don't understand why with does not work:

```
export MAIL_URL=Outlook365://firstname.lastname%40hotmail.com:password@hotmail.com
```
I get error:
```
(Mail not sent; to enable sending, set the MAIL_URL environment variable.)
```
But in beginning of start-wekan.sh I have:
```
echo "MAIL_URL = $MAIL_URL"
```
And it shows that MAIL_URL I did set. Do I need to make some other code changes?

To package.js I did add name and current version for compatibility:
```
Package.describe({
  name: "email",
  summary: "Send email messages",
  version: "2.0.0"
});

//  Actual version. Above 2.0.0 only for compatibility reasons.
//  version: "2.2.0-beta240.1"
```
And then built wekan with `./rebuild-wekan.sh`.

I did also verify that new `email.js` code is at `wekan/.build/bundle/programs/server/packages/email.js`.

**Comment by StorytellerCZ** [comment]
@xet7 You will need the `Package.onUse` section most importantly, but in this case what is messing it up is the custom protocol which then means that URL does not recognize the username and password. I think it might be better to switch to automatically recognize known hosts based on hostname rather than protocol.

**Comment by xet7** [comment]
@StorytellerCZ 

> @xet7 You will need the `Package.onUse` section most importantly, but in this case what is messing it up is the custom protocol which then means that URL does not recognize the username and password. I think it might be better to switch to automatically recognize known hosts based on hostname rather than protocol.

I did use all of new email package, including `package.js` new code.

How would known hosts work? In this case, it's detected from protocol:
```
export MAIL_URL=Outlook365://firstname.lastname%40hotmail.com:password@hotmail.com
```

But usually at O356, Google Workplace, etc companies have custom domain:
```
export MAIL_URL=Outlook365://firstname.lastname%40yourcompany.com:password@yourcompany.com
```

If it were just smtp, it's hard to detect, because `protocol=service` is not mentioned anywhere:
```
export MAIL_URL=smtp://firstname.lastname%40yourcompany.com:password@yourcompany.com
```

Or should service be option?
```
export MAIL_URL=smtp://firstname.lastname%40yourcompany.com:password@yourcompany.com?service=Outlook365
```

**Comment by StorytellerCZ** [comment]
Custom domains would not work with know-services from what I can tell as it only links to the main domain. So in that case it would be useless. Known services takes only login and password and then builds the rest of the url from what it knows.

**Comment by xet7** [comment]
@StorytellerCZ 

With what kind of code I could send email with `Outlook365` service and your newest `email` package code?

**Comment by StorytellerCZ** [comment]
If we look at, for example Gmail, this is what the well-know services knows:

```js
"Gmail": {
        "aliases": ["Google Mail"],
        "domains": ["gmail.com", "googlemail.com"],
        "host": "smtp.gmail.com",
        "port": 465,
        "secure": true
    },
```
So upon detecting Gmail it then sets the host, port and secure connection with the transport and we'll add in the login and password. I'll try to improve the detection of the service a bit more today.
Correct me if I'm wrong, but shouldn't the login for say Gmail also include the custom domain. IE. `storyteller@meteor.com` login send to Gmail services would resolve to the Meteor company account?

**Comment by xet7** [comment]
@StorytellerCZ 

I'll do some tests with different email providers, and then comment here.

**Comment by StorytellerCZ** [comment]
@xet7 Made the improvements to the detection of service. I think it should now cover all the use cases. Let me know if you want to add more tests so that we can be sure about it.

**Comment by StorytellerCZ** [comment]
@xet7 did you get a chance to test the new changes?

**Comment by xet7** [comment]
@StorytellerCZ 

Oh, sorry I did not notice you made new changes. I'll test new changes soon. Thanks!

**Comment by xet7** [comment]
@StorytellerCZ

With your latest changes, with what kind of URL syntax does custom `service` email sending work?

**Comment by xet7** [comment]
@StorytellerCZ

When I tested yesterday, it still did not detect that `MAIL_URL` is set at all.

**Comment by StorytellerCZ** [comment]
The package settings takes precedent. Here is the logic:
```js
const getTransport = function() {
  const packageSettings = Meteor.settings.packages?.email || {};
  // We delay this check until the first call to Email.send, in case someone
  // set process.env.MAIL_URL in startup code. Then we store in a cache until
  // process.env.MAIL_URL changes.
  const url = process.env.MAIL_URL;
  if (this.cacheKey === undefined || (this.cacheKey !== url || this.cacheKey !== packageSettings?.service || 'settings')) {
    if ((packageSettings?.service && wellKnow(packageSettings.service)) || (url && wellKnow(new URL(url).hostname) || wellKnow(url?.split(':')[0] || ''))) {
      this.cacheKey = packageSettings.service || 'settings';
      this.cache = knownHostsTransport(packageSettings, url);
    } else {
      this.cacheKey = url;
      this.cache = url ? makeTransport(url, packageSettings) : null;
    }
  }
  return this.cache;
};
```

**Comment by xet7** [comment]
@StorytellerCZ 

What kind of code is required to send email?

**Comment by xet7** [comment]
It seems to me that `MAIL_URL` is not detected at all.

**Comment by xet7** [comment]
Anyway, I'll test with todays changes, just a moment.

**Comment by StorytellerCZ** [comment]
Did a complete test locally and fixed the bug you were referring to @xet7
Merging now and releasing new 2.4 beta to get more broader testing.

**Comment by xet7** [comment]
@StorytellerCZ 

Thanks a lot! I'll test :)

**Comment by xet7** [comment]
Hmm, interesting error, I'll try to figure it out.

```
> Starting add-description-text-allowed migration.
> Finishing add-description-text-allowed migration.
> Starting add-sort-field-to-boards migration.
> Finishing add-sort-field-to-boards migration.
> Starting add-default-profile-view migration.
> Finishing add-default-profile-view migration.
> Starting add-hide-logo-by-default migration.
> Finishing add-hide-logo-by-default migration.
{"line":"87","file":"percolate_synced-cron.js","message":"SyncedCron: Scheduled \"notification_cleanup\" next run @Fri Jul 30 2021 00:03:08 GMT+0300 (GMT+03:00)","time":{"$date":1627592588955},"level":"info"}
Exception while invoking method 'ATCreateUserServer' TypeError: Cannot read property 'split' of undefined
    at module.exports (/home/wekan/repos/wekan/.build/bundle/programs/server/npm/node_modules/meteor/email/node_modules/nodemailer/lib/well-known/index.js:45:28)
    at knownHostsTransport (packages/email/email.js:81:8)
    at getTransport (packages/email/email.js:107:20)
    at Object.Email.send (packages/email/email.js:221:47)
    at AccountsServer.Accounts.sendVerificationEmail (packages/accounts-password/password_server.js:923:9)
    at MethodInvocation.ATCreateUserServer (packages/useraccounts_core.js:1310:16)
    at packages/check/match.js:118:15
    at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1234:12)
    at Object._failIfArgumentsAreNotAllChecked (packages/check/match.js:116:43)
    at maybeAuditArgumentChecks (packages/ddp-server/livedata_server.js:1768:18)
    at packages/ddp-server/livedata_server.js:719:19
    at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1234:12)
    at packages/ddp-server/livedata_server.js:717:46
    at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1234:12)
    at packages/ddp-server/livedata_server.js:715:46
    at new Promise (<anonymous>)
    at Session.method (packages/ddp-server/livedata_server.js:689:23)
    at packages/ddp-server/livedata_server.js:559:43
```

**Comment by xet7** [comment]
It seems that this line:
```
at module.exports (/home/wekan/repos/wekan/.build/bundle/programs/server/npm/node_modules/meteor/email/node_modules/nodemailer/lib/well-known/index.js:45:28)
```
Has this this kind of split code:
```
module.exports = function (key) {
    key = normalizeKey(key.split('@').pop());
    return normalized[key] || false;
};
```
I try to figure out why it says:
```
Exception while invoking method 'ATCreateUserServer' TypeError: Cannot read property 'split' of undefined
```

**Comment by StorytellerCZ** [comment]
What kind of URL were you using?

**Comment by xet7** [comment]
@StorytellerCZ 

```
export MAIL_URL=Outlook365://firstname.lastname%40hotmail.com:password@hotmail.com

or:

export MAIL_URL=Outlook365://firstname.lastname@hotmail.com:password@hotmail.com
```

**Comment by xet7** [comment]
@StorytellerCZ 

And also:
```
export MAIL_URL=Outlook365://firstname.lastname:password@hotmail.com
```
All have same split error.

**Comment by daxxog** [comment]
Does this change fix [this issue](https://github.com/RocketChat/Rocket.Chat/issues/11846#issuecomment-890418288)? Basically the "name" value needs to be set to the users domain in order for Google Workspace to accept the SMTP EHLO.

**Comment by xet7** [comment]
@daxxog 

Maybe. I think I anyway need to write only nodemailer using code to Wekan, to get this working. Also [Wekan](https://wekan.github.io) is still stuck on Meteor 2.2 and Node 12.x, I have not had any success in upgrading.

**Comment by xet7** [comment]
I'm not saying that any other web framework or programming language is any better than Meteor, everything else I have tried is even more broken in many ways.

**Comment by StorytellerCZ** [comment]
@xet7 I will add your examples to the test suite and fix them, but I'm afraid that since URL can't decipher these objects we will constantly be fixing infinite sets of possible url strings. This though is more of a stop gap measure until people can migrate to the settings option which is going to be 100% accurate.

**Comment by StorytellerCZ** [comment]
@xet7 even with addition of your test cases the tests are still passing for me. I'll do a more complex test later.

**Comment by xet7** [comment]
@StorytellerCZ 

Thanks! I'll test.

**Comment by xet7** [comment]
@StorytellerCZ 

With these examples:
```
export MAIL_URL=Outlook365://firstname.lastname%40hotmail.com:password@hotmail.com
export MAIL_URL=Outlook365://firstname.lastname@hotmail.com:password@hotmail.com
export MAIL_URL=Hotmail://firstname.lastname@hotmail.com:password@hotmail.com
```
I still get this error 
```
> Finishing assign-boardwise-card-numbers migration.
{"line":"87","file":"percolate_synced-cron.js","message":"SyncedCron: Scheduled \"notification_cleanup\" next run @Tue Aug 03 2021 19:56:35 GMT+0300 (GMT+03:00)","time":{"$date":1628009795539},"level":"info"}
(node:15795) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
Exception while invoking method 'ATCreateUserServer' TypeError: Cannot read property 'split' of undefined
    at module.exports (/home/wekan/repos/wekan/.build/bundle/programs/server/npm/node_modules/meteor/email/node_modules/nodemailer/lib/well-known/index.js:45:28)
    at knownHostsTransport (packages/email/email.js:81:8)
    at getTransport (packages/email/email.js:107:20)
    at Object.Email.send (packages/email/email.js:221:47)
    at AccountsServer.Accounts.sendVerificationEmail (packages/accounts-password/password_server.js:923:9)
    at MethodInvocation.ATCreateUserServer (packages/useraccounts_core.js:1310:16)
    at packages/check/match.js:118:15
    at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1234:12)
    at Object._failIfArgumentsAreNotAllChecked (packages/check/match.js:116:43)
    at maybeAuditArgumentChecks (packages/ddp-server/livedata_server.js:1768:18)
    at packages/ddp-server/livedata_server.js:719:19
    at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1234:12)
    at packages/ddp-server/livedata_server.js:717:46
    at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1234:12)
    at packages/ddp-server/livedata_server.js:715:46
    at new Promise (<anonymous>)
    at Session.method (packages/ddp-server/livedata_server.js:689:23)
    at packages/ddp-server/livedata_server.js:559:43
```
That file wekan/.build/bundle/programs/server/npm/node_modules/meteor/email/node_modules/nodemailer/lib/well-known/index.js has this code:
```
/**
 * Resolves SMTP config for given key. Key can be a name (like 'Gmail'), alias (like 'Google Mail') or
 * an email address (like 'test@googlemail.com').
 *
 * @param {String} key [description]
 * @returns {Object} SMTP config or false if not found
 */
module.exports = function (key) {
    key = normalizeKey(key.split('@').pop());
    return normalized[key] || false;
};
```
This nodemailer code still works OK:
```
var nodemailer = require('nodemailer');

let transporter = nodemailer.createTransport({
  service: "Outlook365",
  auth: {
    user: 'firstname.lastname@hotmail.com',
    pass: 'password'
  },
})

let info = transporter.sendMail({
  from: 'Wekan Boards <firstname.lastname@hotmail.com>',
  to: 'me@example.com',
  subject: 'Hi',
  text: 'Hi! Reminder: Meeting tomorrow..',
  html: '<h1>Hi! Reminder: Meeting tomorrow.</h1>'
})
```

**Comment by xet7** [comment]
@StorytellerCZ 

Code I tried with Meteor 2.2 and Node 12.22.4 is at:
https://github.com/wekan/wekan/tree/feature-email-2-2

**Comment by StorytellerCZ** [comment]
@xet7 honestly I don't know what to tell you. I will try a few more things, but so far I was unable to replicate the issue.

**Comment by xet7** [comment]
@StorytellerCZ 

No problem. Please don't use too much time on this. I will fix this in Wekan code anyway.

File: packages/email/email.js [hunk 1]
```diff
@@ -1,11 +1,16 @@
-var Future = Npm.require('fibers/future');
-var urlModule = Npm.require('url');
-var nodemailer = Npm.require('nodemailer');
+import { Meteor } from 'meteor/meteor';
+import { Log } from 'meteor/logging';
+import { Hook } from 'meteor/callback-hook';
 
-Email = {};
-EmailTest = {};
+const Future = Npm.require('fibers/future');
```
Comment (filipenevola) [inline packages/email/email.js:5]: Why not regular imports here?

File: packages/email/email.js [hunk 2]
```diff
@@ -43,27 +48,73 @@ var makeTransport = function (mailUrlString) {
...
+
+  transport._syncSendMail = Meteor.wrapAsync(transport.sendMail, transport);
+  return transport;
+};
+
+// More info: https://nodemailer.com/smtp/well-known/
+const knownHostsTransport = function(settings = undefined, url = undefined) {
+  let service, user, password;
+  if (url && !settings) {
+    let host = url?.split(':')[0];
```
Comment (filipenevola) [inline packages/email/email.js:61]: why do we need ?

File: packages/email/email.js [hunk 3]
```diff
@@ -43,27 +48,73 @@ var makeTransport = function (mailUrlString) {
...
+    const urlObject = new URL(url);
+    if (host === 'http' || 'https') {
```
Comment (filipenevola) [inline packages/email/email.js:63]: this is always going to be true, I believe there is a typo here.

File: packages/email/email.js [hunk 4]
```diff
@@ -145,17 +206,24 @@ Email.hookSend = function (f) {
...
-  var transport = getTransport();
+  let send = true;
+  sendHooks.each(hook => {
+    send = hook(options);
+    return send;
+  });
+  if (!send) return;
+
+  const customTransport = Email.customTransport;
+  const transport = customTransport ? false : getTransport();
```
Comment (filipenevola) [inline packages/email/email.js:221]: too confusing, as the custom transport has priority we should first try to use it, the code will be much simpler.

also, you don't need else's here.

File: packages/email/email.js [hunk 5]
```diff
@@ -145,17 +206,24 @@ Email.hookSend = function (f) {
...
   if (transport) {
     smtpSend(transport, options);
+  } else if (customTransport) {
+    const packageSettings = Meteor.settings.packages?.email || {};
+    customTransport({ settings: { ...packageSettings }, ...options, });
```
Comment (filipenevola) [inline packages/email/email.js:226]: I would call settings as packageSettings as settings would be confusing with Meteor settings.

File: packages/email/email_tests.js [hunk 6]
```diff
@@ -1,14 +1,14 @@
-var streamBuffers = Npm.require('stream-buffers');
+const streamBuffers = Npm.require('stream-buffers');
```
Comment (filipenevola) [inline packages/email/email_tests.js:1]: why not regular imports?

File: packages/email/email_tests.js [hunk 7]
```diff
@@ -83,7 +83,7 @@ Tinytest.add("email - multiple e-mails same stream", function (test) {
       text: "This is the body\nof the message\nFrom us.",
     });
 
-    var contents;
+    let contents;
```
Comment (filipenevola) [inline packages/email/email_tests.js:86]: why not const? We don't need to reuse this variable and change its value

**Review by filipenevola — COMMENTED** [review]
Please see my requested changes.

File: packages/email/email.js (hunk 2 above)
Comment (StorytellerCZ) [reply packages/email/email.js:61]: This is to get known host from url like from this url that @xet7 wants to use: `Outlook365://firstname.lastname:password@hotmail.com`

File: packages/email/email.js (hunk 3 above)
Comment (StorytellerCZ) [reply packages/email/email.js:63]: Not the case if you have url like this: `Outlook365://firstname.lastname:password@hotmail.com`

File: packages/email/email_tests.js (hunk 6 above)
Comment (StorytellerCZ) [reply packages/email/email_tests.js:1]: Fixed

File: packages/email/email.js (hunk 1 above)
Comment (StorytellerCZ) [reply packages/email/email.js:5]: Fixed

File: packages/email/email_tests.js (hunk 7 above)
Comment (StorytellerCZ) [reply packages/email/email_tests.js:86]: Fixed!

File: packages/email/email.js (hunk 5 above)
Comment (StorytellerCZ) [reply packages/email/email.js:226]: Got it! Fixed! Now fixing the docs...

File: packages/email/email.js (hunk 4 above)
Comment (StorytellerCZ) [reply packages/email/email.js:221]: So something like this:
```
const customTransport = Email.customTransport;
  if (customTransport) {
    const packageSettings = Meteor.settings.packages?.email || {};
    customTransport({ packageSettings: { ...packageSettings }, ...options, });
  } else if (Meteor.isProduction) {
    const transport = getTransport();
    smtpSend(transport, options);
  } else {
    devModeSend(options);
  }
```

File: packages/email/email.js (hunk 2 above)
Comment (filipenevola) [reply packages/email/email.js:61]: Ok, but why do we need `?` as you already checked if the url is "true"

File: packages/email/email.js (hunk 3 above)
Comment (filipenevola) [reply packages/email/email.js:63]: I mean, in Javascript this is always going to be "true":

```js
if (host === 'http' || 'https')
```

as 'https' is always "true".

File: packages/email/email.js (hunk 4 above)
Comment (filipenevola) [reply packages/email/email.js:221]: Almost, you don't need to have this variable `customTransport` and also you never need to spread something that is the full content of an object.

```js
{ packageSettings: { ...packageSettings }
```

should be

```js
{ packageSettings,
```

File: packages/email/email.js (hunk 3 above)
Comment (StorytellerCZ) [reply packages/email/email.js:63]: :facepalm:

File: packages/email/email.js (hunk 2 above)
Comment (StorytellerCZ) [reply packages/email/email.js:61]: Remnant from previous version of the code. Fixing...

File: packages/email/email.js (hunk 4 above)
Comment (StorytellerCZ) [reply packages/email/email.js:221]: Fixed

**Pull request merged**